doc/algorithm.md

```diff
@@ -0,0 +1,271 @@
+
+A sensible compromise between texture size and folder size,
+and between simple and complex configuration,
+is to define a [hypergraph](https://en.wikipedia.org/wiki/Hypergraph) on the set of terrains.
+
+### Hypergraph
+
+The terrain hypergraph, for the purposes of Wangscape, is a list of lists of terrains.
+For each list of terrains (which we call cliques), one tileset is generated,
+with a single tile for each combination of four terrains in the clique.
+
+Subject to the restriction that each clique must be unique (ignoring order of terrains),
+the terrain hypergraph can configure the legal terrain combinations as follows:
+* If a set of 2, 3 or 4 terrains is not found in any of the cliques,
+those terrains cannot appear around a common corner.
+In the case of two terrains, this means they cannot be adjacent (even diagonally).
+* If a set of terrains appears in more than one (different) clique,
+extra tiles will be generated for each combination of those tiles.
+The extra tiles will have different borders (except for the single-terrain tiles),
+and can be used to add variety to output displays.
+
+Make sure your hypergraph, where appropriate.
+If your hypergraph is (for instance) `[["a","b"],["c","d"]]`,
+then it will be impossible for terrains `a` and `b` to be on the same map as terrains `c` and `d`.
+
+If you have a set of `N` terrains in which all combinations must be allowed,
+you have a few different options:
+* Put them all in one clique. This will generate the smallest number of tiles,
+but the largest texture size.
+* Choose a clique size `4 <= C < N`, and make one clique for each combination of `C` terrains.
+That's `N!/(C!(N-C)!)` cliques, each with texture size `(rx*C^2, ry*C^2)`.
+Your choice of `C` will depend on your requirements for texture size, folder size, and computation time.
+If `N` is so large that no `C` satisfies your requirements,
+you will need to introduce more adjacency constraints.
+
+## Combining terrains
+
+Each tile has corner terrains `t1, ..., t4`, and is generated as follows:
+
+0. Generate a [corner mask](#generating-corner-masks) for each corner.
+They are essentially 8-bit greyscale images with the same resolution as a tile,
+and should satisfy these constraints:
+    0. The total of the corner masks in each pixel is 255.
+    0. Each one has value 255 at the corresponding corner (hence each one has value 0 at all other corners).
+    0. Along the horizontal border (transitioning from `t1` to `t2` or from `t3` to `t4`) incident on the corresponding corner,
+    each one has values similar to those generated for all other corner masks with a similar horizontal border.
+    0. Equivalent constraints for vertical borders.
+0. Initialise the output tile to opaque black `(0,0,0,255)`.
+0. For each corner mask and corresponding ([rearranged](#rearranging-source-tiles)) terrain texture:
+    0. Add the terrain texture, multiplied by the corner mask and scaled, to the output tile.
```

Comment:
These `0.` subsections are inlined in output document (when you click `View` or go to the branch), you should use `-` or `*` I suppose.